content/blog/2019-11-21-transmeta.md

```diff
@@ -0,0 +1,126 @@
+shows a mean of 33% performance degradation across several applications when
+reordering is disabled.
+
+### Exceptions and Interrupts
+Occasionally compiled code will encounter exceptions in the internal ISA.
+Sometimes these exceptions are the result of speculative compilation (e.g.,
+a reordering of instructions causing a memory fault) but sometimes they are
+genuine exceptions which should be propagated up to the x86 layer (e.g.,
+division by zero).
+
+When Tcached code hits an exception, the CMS issues a rollback instruction to
+restore architectural state to a previous checkpoint and tries interpreting the
+region instead. If the exception goes away, CMS assumes it was due to
+reordering. Otherwise it is a genuine x86 instruction and gets propogated up to
+the program.
+
+Interrupts work similarly to exceptions, but CMS does not try retranslating the
+region in which the interrupt occurs.
+
+### Reordering constraints
+Consider a program that writes a 1 to address `x` and then reads from address
+`y`. If these instructions are reordered and `x = y`, then the resulting
+compiled program will not be correct. Similarly, if `x` and `y` are backed by
+a memory-mapped I/O device, the compiled program would have different I/O
+behavior than the original program.
+
+The CMS turns these potential issues into faults and handles them like
+exceptions. Reorderd instructions are tagged to let the processor know they were
+reordered. Special "alias hardware" does lightweight alias tracking at run time
+and faults if there may be aliasing between two reordered operations.
+Reordered operations that access IO address space also fault. The offending code
+region is then recompiled without the reorderings.
+
+### Self-modifying code
+It is not uncommon for x86 programs to modify themselves. The paper observes
+that it is a standard technique in games, embedded code, and Windows device drivers.
+Following the approach of turning correctness issues into faults, Transmeta
+could (and apparently at one point did) write-protect code pages to cause faults
+and then fall back to interpreting the self-modifying code. Falling back to
+the interpreter is a serious performance penalty for self-modifying programs, so
+the paper includes a few techniques for handling self-modifying code.
+
+Finer-grained write protection can help, since code is likely to be modified
+only in a few places. Crusoe supports this and it gets some speedup over the
+page granularity write protection approach.
+
+Introducing "prologues" that check preconditions on translated code can also
+work. The paper refers to this as self-validation and self-checking. The idea is
+to tack a header onto the translated code which looks up the source page and
+verifies that nothing has changed since it was translated. 
+
+Finally, it is possible to recognize common self-modifying code patterns and
+compile them to ordinary static code. 
```

Comment:
TBH I think this is still the weirdest part of the paper. I am not totally convinced that self-modifying code is important enough to warrant all the ink spilled on making it fast.

It's also somewhat anachronistic… if this paper were written today, everyone would be thinking about the other reason to dynamically generate instructions—proper JITs—and the solution would probably look much different.

Reply:
Yeah, I was confused by the emphasis on self-modifying code. They seemed to have a few techniques for invalidating the Tcache and just kind of glued them together. I'm not sure what would have to change in order to deal with running real JITs either... it seems like the approach might work fine for them if they don't rewrite generated code much.

Reply:
Yeah, that would likely be fine. I was thinking that, in a world with many JITs, a Transmeta-like architecture might try to avoid doing "double profiling," i.e., to let the upper-level JIT hint to the architecture that it has already determined that some generated code is hot—so there's no need to profile it _again_ to rediscover that fact.